scipy/interpolate/_bspl.pyx

```diff
@@ -672,3 +672,173 @@ def evaluate_ndbspline(const double[:, ::1] xi,
                     for i_c in range(num_c_tr):
                         out[j, i_c] = out[j, i_c] + c1r[idx_cflat_base + i_c] * factor
 
+
+@cython.wraparound(False)
+@cython.nonecheck(False)
+@cython.boundscheck(False)
+def _colloc_nd(double[:, ::1] xvals, tuple t not None, npy_int32[::1] k):
+    """Construct the N-D tensor product collocation matrix as a CSR array.
+
+    In the dense representation, each row of the collocation matrix corresponds
+    to a data point and contains non-zero b-spline basis functions which are
+    non-zero at this data point.
+
+    Parameters
+    ----------
+    xvals : ndarray, shape(size, ndim)
+        Data points. ``xvals[j, :]`` gives the ``j``-th data point as an
+        ``ndim``-dimensional array.
+    t : tuple of 1D arrays, length-ndim
+        Tuple of knot vectors
+    k : tuple of ints, length-ndim
+        Spline degrees
+
+    Returns
+    -------
+    csr_data, csr_indices, csr_indptr
+        The collocation matrix in the CSR array format.
+
+    Notes
+    -----
+
+    This routine is in Cython mainly because `find_interval` and `_deBoor_D` are.
```

Comment:
The comment was stale, is removed in the last commit.

Meant to mean "it could have been in pure python, but `find_interval` and `_deBoor_D routines are not exposed to Python, so this routine is also in Cython". But that comment is stale anyway, this routine should stay in Cython, or maybe move down to C at some point in the future.